File: frontend/lib/model/entreprise/banque.dart
import '../pays_model.dart';
import 'type_canaux_paiement.dart';

class BanqueModel {
  final String id;
  final String name;
  final String codeBanque;
  final double soldeReel;
  final double soldeTheorique;
  final PaysModel? country;
  final String? logo;
  final String codeBIC;
  final CanauxPaiement? type;
  final String codeGuichet;
  final String numCompte;
  final String cleRIB;

  BanqueModel({
    required this.id,
    required this.name,
    required this.codeBanque,
    this.country,
    this.type,
    required this.soldeReel,
    required this.soldeTheorique,
    required this.codeBIC,
    this.logo,
    required this.cleRIB,
    required this.numCompte,
    required this.codeGuichet,
  });

  factory BanqueModel.fromJson(Map<String, dynamic> json) {
    return BanqueModel(
      id: json['_id'],
      name: json['name'],
      codeBanque: json['codeBanque'],
      country:
          json['country'] == null ? null : PaysModel.fromJson(json["country"]),
      logo: json['logo'],
      type: json["type"] == null
          ? null
          : canauxPaiementFromString(json["type"]),
      soldeReel:
          json['soldeReel'] == null ? 0 : (json['soldeReel'] as num).toDouble(),
      soldeTheorique: json['soldeTheorique'] == null
          ? 0
          : (json['soldeTheorique'] as num).toDouble(),
      codeBIC: json['codeBIC'] ?? "",
      numCompte: json['numCompte'] ?? "",
      codeGuichet: json['codeGuichet'],
      cleRIB: json['cleRIB'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'codeBanque': codeBanque,
      'soldeTheorique': soldeTheorique.toDouble(),
      'soldeReel': soldeReel.toDouble(),
      'logo': logo,
      'type':
          type == null ? null : canauxPaiementToString(type!),
      'country': country,
      'codeBIC': codeBIC,
      'numCompte': numCompte,
      'cleRIB': cleRIB,
      'codeGuichet': codeGuichet,
    };
  }

  bool equalTo({required BanqueModel bank}) {
    return bank.id == id;
  }
}

File: backend/graphql/resolvers/banque.js
import Banque from "../../models/banque.js";

const banqueModel = new Banque();

const banqueResolvers = {
  banques: async ({ skip, perPage }) =>
    await banqueModel.getAllBanques({
      perPage: perPage,
      skip: skip,
    }),

  banque: async ({ key }) => await banqueModel.getBanque({ key: key }),

  createBanque: async ({
    name,
    codeBanque,
    codeBIC,
    numCompte,
    country,
    type,
    logo,
    codeGuichet,
    cleRIB,
  }) =>
    await banqueModel.createBanque({
      cleRIB: cleRIB,
      codeGuichet: codeGuichet,
      codeBanque: codeBanque,
      name: name,
      type: type,
      codeBIC: codeBIC,
      numCompte: numCompte,
      // soldeReel: soldeReel,
      country: country,
      logo: logo,
    }),

  updateBanque: async ({
    key,
    name,
    codeBIC,
    numCompte,
    codeBanque,
    country,
    type,
    logo,
    codeGuichet,
    cleRIB,
  }) =>
    await banqueModel.updateBanque({
      key: key,
      name: name,
      cleRIB: cleRIB,
      codeBanque: codeBanque,
      codeBIC: codeBIC,
      type: type,
      numCompte: numCompte,
      codeGuichet: codeGuichet,
      logo: logo,
      country: country,
    }),

  // resetBanqueAmount: async ({ key, soldeReel }) =>
  //   await banqueModel.resetBanqueAmount({ key: key, soldeReel: soldeReel, }),

  deleteBanque: async ({ key }) => await banqueModel.deleteBanque({ key: key }),
};

export default banqueResolvers;
